pgmpy/factors/Factor.py

```diff
@@ -496,11 +496,24 @@ def __eq__(self, other):
         if type(self) != type(other):
             return False
 
-        if self.variables == other.variables and all(self.cardinality == other.cardinality) \
-                and all(self.values == other.values):
-            return True
-        else:
+        if set(self.scope()) != set(other.scope()):
             return False
+        else:
+            order_dict = {}
+            for var_index in range(len(self.scope())):
+                order_dict[var_index] = other.scope().index(self.scope()[var_index])
+            if any(self.cardinality[[order_dict[i] for i in range(len(self.cardinality))]] != other.cardinality):
+                return False
+            indexes = np.array(list(itertools.product(*[range(i) for i in self.cardinality])))
```

Comment:
@abinashpanda The problem with `np.fromiter` is that we can't create 2D arrays using it. So, the solution that I could find is to either `chain` our 2D array, use `fromiter` and then reshape it or use structured arrays. With structured arrays we lose indexing functionality because it creates an array of tuples and therefore numpy considers it as a 1D array.

Comparison current method and chain:
```python
%timeit np.array(list(itertools.product(*[range(i) for i in range(2, 10)])))
1 loops, best of 3: 201 ms per loop

%timeit np.fromiter(itertools.chain(*itertools.product(*[range(i) for i in range(2, 10)])), np.int).reshape(np.prod(range(2, 10)), len(range(2, 10)))
10 loops, best of 3: 121 ms per loop

Line #    Mem usage    Increment   Line Contents
================================================
    10   26.410 MiB    0.000 MiB   @profile
    11                             def my_fun2():
    12   26.410 MiB    0.000 MiB       a = itertools.product(*[range(i) for i in range(2, 10)])
    13   65.660 MiB   39.250 MiB       b = list(a)
    14   65.660 MiB    0.000 MiB       c = np.array(a)


Filename: testing.py

Line #    Mem usage    Increment   Line Contents
================================================
     4   23.352 MiB    0.000 MiB   @profile
     5                             def my_fun():
     6   23.352 MiB    0.000 MiB       a = itertools.product(*[range(i) for i in range(2, 10)])
     7   66.180 MiB   42.828 MiB       b = itertools.chain(*a)
     8   26.406 MiB  -39.773 MiB       np.fromiter(b, np.int).reshape(np.prod(range(2, 10)), len(range(2, 10)))
```
So creating a list is slower and using `chain` is using almost the same amount of memory as creating a list but we are losing readability. So, which method do you think we should go with?